### PR TITLE
Full tinydns spec-compliant source implementation

### PR DESCRIPTION
**Pull request by ross**
Completely rewrites the TinyDNS source to be fully/completely spec compliant for all of the base types and IPv6 extensions. Specifically handles all the auto-creation of `NS`/`MX` `A`s and such that previously wasn't implemented.

For the most part the original unit tests haven't changed with the exception of now expecting the auto-created As and exercising more of the edge cases: trailing dots, w/wo IPs etc.

Won't be surprised if there's problems/details in here, but it should be in at least as good a shape now as it was before and much more complete/powerful. 

/cc https://github.com/octodns/octodns/issues/999
/cc @dbuenoparedes

### CODE REVIEW

**Comment by viranch** [comment]
Not really familiar with tinydns to review this. Happy to give a 🚢 to make CI happy.

**Comment by ross** [comment]
> Not really familiar with tinydns to review this

That's fine, more just looking for a set of :eyes: to pass over it and make sure that nothing is 🙃 or doesn't make any sense to people who aren't me :grin:

**Review by viranch — APPROVED** [review]
There seems to be some room for minimizing repetitive code or turning it into method(s), specific instances below. Although edge cases may be preventing some/all of it or making it not worth the work involved.

Otherwise/code-wise looks okay to me!

File: octodns/source/tinydns.py [hunk 1]
```diff
@@ -3,104 +3,459 @@
...
+        # All record types, including those registered by 3rd party modules
+        return set(Record.registered_types().keys())
+
+    def _records_for_at(self, zone, name, lines, arpa=False):
+        # @fqdn:ip:x:dist:ttl:timestamp:lo
+        # MX (and optional A)
+
+        if arpa:
+            # no arpa
+            return []
```
[the remark this answers is not in the document]
Comment (ross) [reply octodns/source/tinydns.py:46]: That'd require the caller to know what the various symbol (`_records_for_*`) methods do with arpa. Some only process arpa=True, others arpa=False, and a couple call other `_records_for_*`, one for each case.

File: octodns/source/tinydns.py [hunk 2]
```diff
@@ -3,104 +3,459 @@
...
+
+        if not zone.owns('MX', name):
+            # if name doesn't live under our zone there's nothing for us to do
+            return
```
[the remark this answers is not in the document]
Comment (ross) [reply octodns/source/tinydns.py:50]: Similar to the `arpa` checks above there are subtle differences in what's checked based on the specifics of the `_records_for_*` function, most of the forward stuff do this simple check. Things like `NS`, `MX`, `SRV`, have to check the `A`s they're auto-creating when there's IPs to make sure they live in the current zone. Similarly for `PTR` we don't know the FQDN to check until after we create the IP and get it's reverse name.

File: octodns/source/tinydns.py [hunk 3]
```diff
@@ -3,104 +3,459 @@
...
+
+        # see if we can find a ttl on any of the lines, first one wins
+        ttl = self.default_ttl
+        for line in lines:
+            try:
+                ttl = int(line[4])
+                break
+            except IndexError:
+                pass
```
[the remark this answers is not in the document]
Comment (ross) [reply octodns/source/tinydns.py:59]: Yeah. I thought of that while writing one of them, but apparently the thought didn't stick long enough to actually do it.
